### PR TITLE
Bucket.search

### PR DESCRIPTION
**Pull request by meffij**
Tests in progress

### CODE REVIEW

File: api/python/t4/bucket.py [hunk 1]
```diff
@@ -27,6 +39,47 @@ def __init__(self, bucket_uri):
...
+            version_id: version_id of object version
+            operation: Create or Delete
+            meta: metadata attached to object
+            size: size of object in bytes
+            text: indexed text of object
+            source: source document for object (what is actually stored in ElasticSeach)
+            time: timestamp for operation
+
+        """
+        return search(query, self._search_endpoint)
```
Comment (ResidentMario) [inline api/python/t4/bucket.py]: We're silently placing the user in a scenario where the bucket they are interacting with may or may not be searchable. If they later go to try to search that bucket, and that bucket isn't searchable, they get some obscure error---frustrating.

We should throw a better error when search fails because the bucket isn't properly configured for it.

File: api/python/t4/search_util.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                'meta': meta,
+                'size': size,
+                'text': text,
+                'source': result['_source'],
+                'time': time
+            })
+        results = list(sorted(results, key=lambda x: x['time'], reverse=True))
+        return results
+    except KeyError:
+        return raw_response
```
Comment (ResidentMario) [inline api/python/t4/search_util.py]: Why would returning a raw response object in this case be useful to the end user? If my function fails I'd generally expect to get an error back, not a blob.

File: api/python/tests/test_search.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                        'type': 'asdf',
+                        'user_meta': {},
+                        'size': 0,
+                        'text': ''
+                    }
+                }]
+            }
+        }
+        create_es_mock.return_value = es_mock
+        b = Bucket('s3://quilt-example')
```
Comment (ResidentMario) [inline api/python/tests/test_search.py]: We use `s3://test-bucket` as our mock bucket elsewhere, should do here too for consistency.

File: api/python/t4/search_util.py (hunk 2 above)
Comment (meffij) [reply api/python/t4/search_util.py]: no good reason; it's more helpful to the user to see whatever Elasticsearch says their problem was than to see a generic error from us

Comment (ResidentMario) [reply api/python/t4/search_util.py]: I doubt most users will even know what an Elasticsearch is, let alone that that's what we're using, much better here to be consistent and up-front about the error.

File: api/python/t4/bucket.py (hunk 1 above)
Comment (meffij) [reply api/python/t4/bucket.py]: I made the error messages a little better, what do you think?

File: api/python/t4/search_util.py (hunk 2 above)
Comment (meffij) [reply api/python/t4/search_util.py]: What verbiage would you recommend?

File: api/python/t4/bucket.py (hunk 1 above)
Comment (ResidentMario) [reply api/python/t4/bucket.py]: Current error message:

> QuiltException: Config info not found for this bucket at config_url

I would shorten this to just:

> QuiltException: Config info not found for this bucket

"at `config_url`" implies to me that there is `Bucket.config_url` attribute that I can access to further debug the issue, when in reality no such attribute exists, which is unhelpful.

Once we have documentation we'll need a section for FAQs or common issues that contains an entry for this config problem, which we'll probably want to point the user to from this error message. Task for the future.

File: api/python/t4/search_util.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    raw_response = es_client.search(index=ES_INDEX, body=payload)
+
+    try:
+        results = []
+        for result in raw_response['hits']['hits']:
+            key = result['_source']['key']
+            vid = result['_source']['version_id']
+            operation = result['_source']['type']
+            meta = json.dumps(result['_source']['user_meta'])
```
Comment (stevededalus) [inline api/python/t4/search_util.py]: Is this field guaranteed to be set?

File: api/python/t4/search_util.py (hunk 2 above)
Comment (stevededalus) [reply api/python/t4/search_util.py]: Seems like it should raise an exception that contains an error message with the raw response inside it?

Comment (ResidentMario) [reply api/python/t4/search_util.py]: I'd throw `QuiltError("Query failed unexpectedly due to either a bad query or a misconfigured search service.")` (or similar).

Comment (ResidentMario) [reply api/python/t4/search_util.py]: @stevededalus I've never patched an `Exception` to provide a failing response as a param (is that what you are suggesting?), though in retrospect [obviously](https://stackoverflow.com/questions/16466406/passing-an-object-with-an-exception) you can do that. Is this something that one would reasonably expect from an API?

Comment (stevededalus) [reply api/python/t4/search_util.py]: No, but it's useful in `alpha` for us to debug customer issues.

File: api/python/t4/search_util.py (hunk 4 above)
Comment (meffij) [reply api/python/t4/search_util.py]: yes it will be `{}` if the indexer doesn't see any user meta

File: api/python/t4/search_util.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+def search(query, search_endpoint, aws_region='us-east-1'):
+    """
+    Searches your bucket. query can contain plaintext, and can also contain clauses
+    like $key:"$value" that search for exact matches on specific keys.
+
+    Returns either the request object (in case of an error)
+            or a list of objects with the following keys:
+        key: key of the object
+        version_id: version_id of object version
+        operation: Create or Delete
```
[the remark this answers is not in the document]
Comment (meffij) [reply api/python/t4/search_util.py]: It's either `"Create"` or `"Delete"` -- when you delete something, we insert a document into ES with barebones data and the operation type "Delete"

[the remark this answers is not in the document]
Comment (stevededalus) [reply api/python/t4/search_util.py]: So this is a tombstone object that signifies that the entry is deleted?
